### DISCUSSION TITLE
Display Past Generation Times in the Console

### DISCUSSION THREAD

**Discussion by koinkoin-project** [Woolverine94/biniou#4] [Ideas]
Hello

Advantages:

Avoid staying glued to the screen to ensure you don't miss the end of generations and can thus record the time of the much-awaited generation :)

**Comment by Woolverine94** [comment]
I agree. biniou definitely lacks this features, which will be very useful to produce metrics when testing new optimizations and confirm benefits of them. 

I add this idea to the to-do-list.

**Reply by koinkoin-project** [reply]
I hadn't seen this, but the time information seems to be already displayed in the console,
for exemple,
==> llama_print_timings:       total time =  318059.21 ms

**Reply by Woolverine94** [reply]
In the case of the chatbot, it's most likely the output of llama-cpp-python. 

Modules based on diffusers / transformers (i.e. : most of them) doesn't display this information.

I've already implemented a working timer, but i'm trying now to extend this to a generic "metric" function that could send (in the future) other informations than generation time.

**Reply by koinkoin-project** [reply]
> ...which could (in future) send information other than generation time.

You've caught my attention: roughly what kind of information, and what about the output (console, text file)?

**Reply by Woolverine94** [reply]
Still don't have any precise idea, only "provisioning" this possibility. 
Could be (maybe) RAM consumption, or any metrics that can be measured before and after inference. 
Output could be console only, as it's easy to create a txt log file based on the console output.

**Reply by koinkoin-project** [reply]
I was thinking in the background and this came to mind:

- Record time in milliseconds (already implemented) + add time information for humans (e.g. 1 hour 54 minutes or 1 hour 11 minutes 34 seconds :)
- Write the prompt and as many options as possible in a txt file with the same name as the output file provided by biniou in /output)
- Save the file in the gradio browser (with a clickable link to launch an additional information file)
- Save the output file (with a clickable link to launch an image or video or txt file)

**Reply by Woolverine94** [reply]
> Record time in milliseconds (already implemented) + add time information for humans (e.g. 1 hour 54 minutes or 1 hour 11 minutes 34 seconds :)

Why not, but not a priority. Seconds will be simpler to handle for benchmarks and comparisons. 

> Write the prompt and as many options as possible in a txt file with the same name as the output file provided by biniou in /output)
Save the file in the gradio browser (with a clickable link to launch an additional information file)
Save the output file (with a clickable link to launch an image or video or txt file)

Duplicates of console log features introduced in commit [c7f10a5](https://github.com/Woolverine94/biniou/commit/c7f10a59519438739ac594e4c21302b11bc76a02), pursued through commit [535da5c](https://github.com/Woolverine94/biniou/commit/535da5c4f7b6df1754c22486b540129d2cede75a) and which has to be finished for text, audio and 3d modules. The advantage of using the log output instead of txt files per generation is keeping track of the workflow through the modules.

**Comment by Woolverine94** [comment]
Commit [da5c752](https://github.com/Woolverine94/biniou/commit/da5c7522039400b923cd8f7f95cc7a9cf9a63771) introduce this functionality for all modules.

**Reply by koinkoin-project** [reply]
Hello,

Yes I saw the update from biniou you were quick. I'll test to see and run a series of cpu/gpu processes that I'll try to classify by models tested. And try to post the endless terminal output as message attachments so as not to make it difficult to read the information displayed.

I'll start very soon,

**Reply by Woolverine94** [reply]
Hello @koinkoin-project , 

Don't rush. I'm currently working on the logs features and will try to integrate the console output directly inside biniou webui, which will be more comfortable to use.

**Reply by Woolverine94** [reply]
Commit [4d2fe79](https://github.com/Woolverine94/biniou/commit/4d2fe797d99b5289285193c02635bd063fd89ecc) introduce a minimal console for all modules.

**Reply by koinkoin-project** [reply]
hello, look :

`Traceback (most recent call last):
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/gradio/queueing.py", line 407, in call_prediction
    output = await route_utils.call_process_api(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/gradio/route_utils.py", line 226, in call_process_api
    output = await app.get_blocks().process_api(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/gradio/blocks.py", line 1550, in process_api
    result = await self.call_function(
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/gradio/blocks.py", line 1185, in call_function
    prediction = await anyio.to_thread.run_sync(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/gradio/utils.py", line 661, in wrapper
    response = f(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/env/lib/python3.11/site-packages/gradio/utils.py", line 661, in wrapper
    response = f(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/ressources/txt2img_sd.py", line 78, in image_txt2img_sd
    nsfw_filter_final, feat_ex = safety_checker_sd(model_path_txt2img_sd, device_txt2img_sd, nsfw_filter)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/biniou/biniou/ressources/common.py", line 210, in safety_checker_sd
    print("Date et heure actuelles : ", datetime.datetime.now())
                                        ^^^^^^^^
NameError: name 'datetime' is not defined
`

+ gradio error

**Reply by Woolverine94** [reply]
Hello @koinkoin-project,  

> File "/home/biniou/biniou/ressources/common.py", line 210, in safety_checker_sd print("Date et heure actuelles : ", datetime.datetime.now()) ^^^^^^^^ NameError: name 'datetime' is not defined 

This is not my code ...

**Reply by koinkoin-project** [reply]
> This is not my code ...

LOL, but I'd just done a Biniou update, weird there really are questions I wouldn't have answers to. It seemed to me that the Biniou updates were de facto overwriting my modifications. And I even had the impression that this was the second update that updated a lot of files (after deleting my local common.py file).

**Reply by Woolverine94** [reply]
The first step of the update script `git pull` the repo and does not use any kind of merging strategies. You should `git checkout .` before running the update script if you have made modifications ... and made them again after update. At this time, biniou is not thinked at all to merge users modifications within his code. 

Anyway, even if I can't give you support on custom modifications of biniou (mostly because I do not have access to your code), it seems the error you encountered in the previous error message is a missing `from datetime import datetime` at the top of common.py.

**Reply by koinkoin-project** [reply]
> Anyway, even if I can't give you support on custom modifications of biniou (mostly because I do not have access to your code), it seems the error you encountered in the previous error message is a missing from datetime import datetime at the top of common.py.

Don't worry, it was some badly screwed up time overlay tests in the console, and in a function where it seemed to me that all models passed... Your code is much better than mine :)

> The first step of the update script git pull the repo and does not use any kind of merging strategies. You should git checkout . before running the update script if you have made modifications ... and made them again after update. At this time, biniou is not thinked at all to merge users modifications within his code.

I don't quite understand: if I make changes to the resources/common.py file tomorrow, for example, then the day after tomorrow I'll do a biniou update and my modified common.py won't be updated? As I write these lines, I know that I'm going to test more precisely to memorize the behavior and real impact of update.sh on my modifications to this common.py file.

**Reply by Woolverine94** [reply]
>Don't worry, it was some badly screwed up time overlay tests in the console, and in a function where it seemed to me that all models passed... Your code is much better than mine :)

I can confirm that this function is ***NOT*** common to all models :D 

> I don't quite understand: if I make changes to the resources/common.py file tomorrow, for example, then the day after tomorrow I'll do a biniou update and my modified common.py won't be updated?

Yes, it won't be updated until you cancel your changes. There's a lot of informations in the log output while updating, but the first lines should mention that the `git pull` has failed because of changes you introduced (which is not a good thing at all, because the following update commands could needs the new revision of the files). 

You should do a `git checkout .` to revoke your changes, followed by a `git status` to validate you're on a clean copy ***before*** using update.sh. Maybe there is also something to try with the `autostash` option of git pull, but i have no experience of it. 

Anyway, biniou is not at this point designed to be modified, then updated with a merge of the changes and probably should never be.
